### PR TITLE
feat: verify is tx is confirmed before getting nc state

### PR DESCRIPTION
**Pull request by pedroferreira1**
### Acceptance Criteria
- Verify if transaction is confirmed before getting nc state and history. If it's not confirmed, add a loading and keep polling the data.

The currently screen when we get the state before the tx is confirmed

image

New screen waiting confirmation

image

### Security Checklist
- [x] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/screens/nano-contract/NanoContractDetail.js [hunk 1]
```diff
@@ -116,8 +121,55 @@ function NanoContractDetail() {
...
+  /**
+   * Validates if the transaction is a valid nano contract creation tx
+   * and if it's confirmed by a block already.
+   *
+   * If everything is ok, calls loadData, otherwise calls itself again
+   * in 5s if it's waiting for a block confirmation
+   *
+   * We need this because the state and history are only loaded when the tx is confirmed
+   */
+  const validateAndLoad = async () => {
```
Comment (andreabadesso) [inline src/screens/nano-contract/NanoContractDetail.js:133]: _Comment unrelated to line number_

Should we stop trying at some point?

File: src/screens/nano-contract/NanoContractDetail.js [hunk 2]
```diff
@@ -116,8 +121,55 @@ function NanoContractDetail() {
...
+      setErrorMessage(t`Transaction must be a nano contract creation.`);
+      setLoading(false);
+      return;
+    }
+
+    const isConfirmed = response.meta.first_block !== null;
+    if (!isConfirmed) {
+      // Wait for transaction to be confirmed
+      setWaitingConfirmation(true);
+      setTimeout(validateAndLoad, 5000);
```
Comment (andreabadesso) [inline src/screens/nano-contract/NanoContractDetail.js:163]: I think you should cancel this timeout when the screen is unmounted

In the current implementation it might run forever in the background until the tx is confirmed

File: src/screens/nano-contract/NanoContractDetail.js (hunk 1 above)
Comment (pedroferreira1) [reply src/screens/nano-contract/NanoContractDetail.js]: The network might take some minutes to confirm the transaction. I think it's ok to query the full node once every 5s to get this information if the user is still with the screen opened.

File: src/screens/nano-contract/NanoContractDetail.js (hunk 2 above)
Comment (pedroferreira1) [reply src/screens/nano-contract/NanoContractDetail.js]: Thanks, it was leaking. I removed the `setTimeout` with the latest refactor. Can you see it again, please?

File: src/constants.js [hunk 3]
```diff
@@ -291,3 +291,22 @@ export const NANO_UPDATE_ADDRESS_LIST_COUNT = 5;
  * Number of elements in the list of transactions of a nano contract
  */
 export const NANO_CONTRACT_HISTORY_COUNT = 5;
+
+/**
+ * Base statuses for saga reducer handlers
+ * Used by all other statuses objects
+ */
+const SAGA_BASE_STATUS = {
```
Comment (andreabadesso) [inline src/constants.js:299]: I think you can just call it `BASE_STATUS`

File: src/sagas/nanoContract.js [hunk 4]
```diff
@@ -124,9 +128,108 @@ export function* updateNanoContractRegisteredAddress({ payload }) {
...
+  if (!isConfirmed) {
+    // Wait for transaction to be confirmed
+    yield put(nanoContractDetailSetStatus({ status: NANO_CONTRACT_DETAIL_STATUS.WAITING_TX_CONFIRMATION }));
+
+    // This confirmation might take some minutes, depending on network mining power
+    // then it's ok to wait until it's confirmed, or until the user gives up and
+    // leaves the screen
+    while (true) {
+      // Wait 5s, then we fetch the data again to check if is has been confirmed
+      yield delay(5000);
```
Comment (andreabadesso) [inline src/sagas/nanoContract.js:176]: Maybe use a constant?

File: src/actions/index.js [hunk 5]
```diff
@@ -63,6 +63,9 @@ export const types = {
   NANOCONTRACT_EDIT_ADDRESS: 'NANOCONTRACT_EDIT_ADDRESS',
   NANOCONTRACT_UNREGISTER: 'NANOCONTRACT_UNREGISTER',
   BLUEPRINT_ADD_INFORMATION: 'BLUEPRINT_ADD_INFORMATION',
+  NANOCONTRACT_DETAIL_REQUEST: 'NANO_CONTRACT_DETAIL_REQUEST',
+  NANOCONTRACT_DETAIL_SET_STATUS: 'NANO_CONTRACT_DETAIL_SET_STATUS',
+  NANOCONTRACT_DETAIL_LOADED: 'NANO_CONTRACT_DETAIL_LOADED',
```
Comment (andreabadesso) [inline src/actions/index.js:68]:
```suggestion
  NANOCONTRACT_LOAD_DETAILS_REQUESTED: 'NANOCONTRACT_LOAD_DETAILS_REQUESTED',
  NANOCONTRACT_LOAD_DETAILS_STATUS_UPDATE: 'NANOCONTRACT_LOAD_DETAILS_STATUS_UPDATE',
  NANOCONTRACT_LOAD_DETAILS_SUCCESS: 'NANOCONTRACT_LOAD_DETAILS_SUCCESS',
```

File: src/screens/nano-contract/NanoContractDetail.js [hunk 6]
```diff
@@ -51,25 +53,57 @@ function NanoContractDetail() {
...
   const nc = nanoContracts[ncId];
-  let blueprintInformationAux = nc != null ? blueprintsData[nc.blueprintId] : null;
 
   // loading {boolean} Bool to show/hide loading element
   const [loading, setLoading] = useState(true);
   // data {Object} Nano contract loaded data
   const [data, setData] = useState(null);
   // blueprintInformation {Object} Blueprint information data
-  const [blueprintInformation, setBlueprintInformation] = useState(blueprintInformationAux);
+  const [blueprintInformation, setBlueprintInformation] = useState(null);
```
Comment (andreabadesso) [inline src/screens/nano-contract/NanoContractDetail.js:62]: Update docs right above this to `Object | null`

File: src/screens/nano-contract/NanoContractDetail.js [hunk 7]
```diff
@@ -51,25 +53,57 @@ function NanoContractDetail() {
...
 
+  useEffect(() => {
+    if (nanoContractDetailState.status === NANO_CONTRACT_DETAIL_STATUS.LOADING) {
+      setLoading(true);
+    }
+
+    if (nanoContractDetailState.status === NANO_CONTRACT_DETAIL_STATUS.ERROR) {
+      setLoading(false);
+      setErrorMessage(nanoContractDetailState.error);
+    }
```
Comment (andreabadesso) [inline src/screens/nano-contract/NanoContractDetail.js:89]: Do you really need this? 

Can't you just use `nanoContractDetailsState` directly when rendering?

File: src/screens/nano-contract/NanoContractDetail.js [hunk 8]
```diff
@@ -116,49 +150,15 @@ function NanoContractDetail() {
...
-      const state = await hathorLib.ncApi.getNanoContractState(ncId, Object.keys(blueprintInformationAux.attributes), ['__all__'], []);
-      setData(state);
-    } catch(e) {
-      // Error in request
-      setErrorMessage(t`Error getting nano contract state.`);
-    };
-  }
-
   const renderBody = () => {
     if (loading) {
```
Comment (andreabadesso) [inline src/screens/nano-contract/NanoContractDetail.js:154]:
```suggestion
    if (nanoContractDetailState.status === NANO_CONTRACT_DETAIL_STATUS.LOADING) {
```

File: src/screens/nano-contract/NanoContractDetail.js (hunk 7 above)
Comment (andreabadesso) [reply src/screens/nano-contract/NanoContractDetail.js:89]: I think we should always avoid having stateful components

The state is our enemy

**Review by tuliomir — APPROVED** [review]

File: src/actions/index.js [hunk 9]
```diff
@@ -614,3 +617,35 @@ export const nanoContractUnregister = (ncId) => ({
...
+  type: types.NANOCONTRACT_DETAIL_SET_STATUS,
+  payload: { status, error },
+});
+
+/**
+ * Set nano contract detail state in redux
+ *
+ * @param {hathorLib.nano_contracts.types.NanoContractStateAPIResponse} state
+ */
+export const nanoContractDetailLoaded = (state) => ({
```
[the remark this answers is not in the document]
Comment (pedroferreira1) [reply src/actions/index.js]: Done https://github.com/HathorNetwork/hathor-wallet/pull/695/commits/a6bb889b4f8c792312cd029a9ddaec70c5039564

File: src/actions/index.js (hunk 5 above)
Comment (pedroferreira1) [reply src/actions/index.js]: Done https://github.com/HathorNetwork/hathor-wallet/pull/695/commits/561eca6fb5f18ddba6f2891dbd1c469b75ab7d2d

File: src/constants.js (hunk 3 above)
Comment (pedroferreira1) [reply src/constants.js]: Done https://github.com/HathorNetwork/hathor-wallet/pull/695/commits/cedc8b79595c55ce8dcdac5881518afb64b75ae8

File: src/sagas/nanoContract.js (hunk 4 above)
Comment (pedroferreira1) [reply src/sagas/nanoContract.js]: Done https://github.com/HathorNetwork/hathor-wallet/pull/695/commits/67b0aa8583baad15a93939454a4ad5e148035fcb

File: src/screens/nano-contract/NanoContractDetail.js (hunk 6 above)
Comment (pedroferreira1) [reply src/screens/nano-contract/NanoContractDetail.js:60]: Done https://github.com/HathorNetwork/hathor-wallet/pull/695/commits/70c70274e815e718dff72325557821dc65876b25

**Review by andreabadesso — APPROVED** [review]

**Review by andreabadesso — APPROVED** [review]
